scipy/signal/spline.py

```diff
@@ -0,0 +1,41 @@
+# THIS NAMESPACE IS NOT AND NEVER WAS PUBLIC - PLEASE DO NOT USE IT!
+
+import warnings
+
+from . import _spline
+
+
+def cspline2d(input, lambda=0.0, precision=-1.0):
+    warnings.warn("Please use `cspline2d` from the `scipy.signal` namespace, "
+                  "the `scipy.signal.spline` namespace is deprecated.",
+                  category=DeprecationWarning)
+    return _spline.cspline2d(input, lambda=lambda, precision=precision)
```

Comment:
How about using a module level `__getattr__`:

```python
import warnings
from . import _spline

__all__ = ['cspline2d', 'qspline2d', 'sepfir2d', 'symiirorder1', 'symiirorder2']


def __dir__():
    return __all__


def __getattr__(name):
    if name not in __all__:
        raise AttributeError(
            f"scipy.signal.spline is deprecated and has no attribute {name}. "
            "Try looking in scipy.signal instead.")

    warnings.warn(f"Please use `{name}` from the `scipy.signal` namespace, "
                  "the `scipy.signal.spline` namespace is deprecated.",
                  category=DeprecationWarning, stacklevel=2)
    return getattr(_spline, name)

```

This saves some boiler plate code, and also makes `from x import foo` warn as well as calls through attribute access like `spline.foo` which isn't necessarily where the function is called.  So, the stack trace will point directly to the site that needs to be changed.

It also reveals that `bspline.py` needs to be updated to import from the underscored module just after importing `scipy` and without needing tests to run.

Reply:
Nice idea! Agreed that that is more maintainable. It also avoids having to fiddle with docstrings.